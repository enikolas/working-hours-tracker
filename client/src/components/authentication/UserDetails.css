.user-details {
  position: absolute;
  top: 0;
  right: 20px;
  background: #5a717f;
  text-align: center;
  overflow: hidden;
  height: 29px;
  transition: all ease-in 0.5s;
}

.user-details:hover {
  height: 102px;
}

.user-details .employee-name {
  padding: 10px;
  font-weight: bold;
  color: #c5ccd1;
  font-size: 12px;
  min-width: 130px;
}

.user-details button {
  background: transparent;
  color: #fff;
  cursor: pointer;
  opacity: 0.7;
}

.user-details button:hover {
  opacity: 1;
}

.user-details button.logout {
  background-color: #909fa8;
  text-transform: lowercase;
  padding: 2px 6px;
  color: #fff;
  cursor: pointer;
  margin-top: 4px;
}

.user-details button.logout:hover {
  background-color: #c5ccd1;
}

.user-details button.logout:focus {
  outline: 1px solid #f93;
}

.user-details nav li {
  list-style: none;
}
