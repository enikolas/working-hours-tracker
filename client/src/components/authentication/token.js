const API_AUTH_TOKEN = 'achiever-auth-token';

export const setAuthToken = token => localStorage.setItem(API_AUTH_TOKEN, token);

export const getAuthToken = () => localStorage.getItem(API_AUTH_TOKEN);

export const removeAuthToken = () => localStorage.removeItem(API_AUTH_TOKEN);
