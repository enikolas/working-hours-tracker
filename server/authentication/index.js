require('dotenv').config();

const HEADER_REGEX = /Bearer (.*)$/;

exports.authenticate = ({ req }) => {
  const authorization = req.headers.authorization || '';
	const authorizationHeader = HEADER_REGEX.exec(authorization);

	if (!authorizationHeader) {
		return null;
	}

	const token = authorizationHeader[1];

	return token;
};
