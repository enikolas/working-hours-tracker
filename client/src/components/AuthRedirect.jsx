import React from 'react';
import PropTypes from 'prop-types';
import { graphql } from '@apollo/client/react/hoc';

import * as queries from '../queries';
import FullScreenSpinner from './ui/FullScreenSpinner';
import Login from './Login';
import TimeEntry from '../containers/TimeEntryContainer';
import { getAuthToken, removeAuthToken } from './authentication/token';

const _checkAuth = () => Boolean(getAuthToken());

class AuthRedirect extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			authenticated: null
		};
	}

	componentWillMount() {
		this.setState({ authenticated: _checkAuth() });
	}

	componentWillReceiveProps(nextProps) {
		const { userDetails, loading, error } = nextProps.userDetailsQuery;
		if (error || (!loading && !userDetails)) {
			removeAuthToken();
			this.setState({ authenticated: false });
		} else {
			this.setState({ authenticated: _checkAuth() });
		}
	}

	render() {
		if (this.props.userDetailsQuery.loading) {
			return <FullScreenSpinner active={this.props.userDetailsQuery.loading} />;
		}
		const Component = this.state.authenticated ? TimeEntry : Login;
		return Component ? <Component /> : null;
	}
}

AuthRedirect.propTypes = {
	userDetailsQuery: PropTypes.object
};

AuthRedirect.defaultProps = {
	userDetailsQuery: {}
};

export default graphql(queries.userDetails, { name: 'userDetailsQuery' })(AuthRedirect);
