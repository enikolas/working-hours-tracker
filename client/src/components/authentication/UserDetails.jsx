import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql } from '@apollo/client/react/hoc';
import { compose } from 'recompose';

import * as queries from '../../queries';
import strings from '../../utils/strings';
import { getAuthToken, removeAuthToken } from './token';
import { clearTodayStorage } from '../../utils';
import UserOptions from '../userOptions/UserOptions';

import './UserDetails.css';

const _logout = (event) => {
	event.preventDefault();
	removeAuthToken();
	clearTodayStorage();
	window.location.reload();
};

class UserDetails extends Component {
	constructor(props) {
		super(props);

		this.state = {
			authenticated: false
		};

		this.openMonthlyReport = this.openMonthlyReport.bind(this);
		this.changePassword = this.changePassword.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		const { userDetails, loading, error } = nextProps.userDetailsQuery;

		if (error || (!loading && !userDetails)) {
			removeAuthToken();
			this.setState({ authenticated: false });
		} else {
			const token = getAuthToken();
			const authenticated = Boolean(token && userDetails.name);
			this.setState({ authenticated });
		}
	}

	/* eslint-disable class-methods-use-this */

	openMonthlyReport(monthData) {
		console.log({ monthData });
	}

	async changePassword(passwordData) {
		const currentPassword = passwordData.old;
		const newPassword = passwordData.new;

		const response = await this.props.changePassword({
			variables: {
				currentPassword,
				newPassword
			}
		});

		if (response) {
			if (response.errors && response.errors.length) {
				// TODO: shows error message
			} else {
				removeAuthToken();
				window.location.reload();
			}
		}
	}

	/* eslint-enable class-methods-use-this */

	render() {
		const { userDetails, loading, error } = this.props.userDetailsQuery;

		if (loading) {
			return <div>Loading...</div>;
		}

		const { authenticated } = this.state;

		if (authenticated && !error) {
			const [firstName] = userDetails.name.split(' ');

			return (
				<div className="user-details">
					<div className="employee-name">
						{`${strings.helloName} ${firstName}`}
					</div>
					<nav>
						<UserOptions
							reportCall={this.openMonthlyReport}
							passwordCall={this.changePassword}
						/>
						<li>
							<button className="logout" onClick={_logout}>
								{strings.logout}
							</button>
						</li>
					</nav>
				</div>
			);
		}
		return null;
	}
}

export default compose(
	graphql(queries.userDetails, { name: 'userDetailsQuery' }),
	graphql(queries.changePassword, { name: 'changePassword' })
)(UserDetails);

UserDetails.propTypes = {
	userDetailsQuery: PropTypes.object,
	changePassword: PropTypes.func
};

UserDetails.defaultProps = {
	userDetailsQuery: {},
	changePassword: () => {}
};
