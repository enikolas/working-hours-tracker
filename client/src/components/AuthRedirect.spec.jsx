import React from 'react';
import { shallow } from 'enzyme';

import AuthRedirect from './AuthRedirect';

import * as token from './authentication/token';

const defaultUserDetailsQuery = {
	loading: false,
	userDetails: {
		name: 'JOHN DOE',
		dailyContractedHours: '8:00',
		lastFridayBalance: ' 2:20',
		__typename: 'UserDetails'
	},
	error: false
};

const shallowComponent = (userDetailsQuery) => {
	const wrapper = shallow(<AuthRedirect
		userDetailsQuery={userDetailsQuery}
		signIn={{}}
		addTimeEntryMutation={() => {}}
		updateTimeEntryMutation={() => {}}
	/>);
	return wrapper;
};

describe('AuthRedirect', () => {
	describe('render', () => {
		it('should render page loading if loading', () => {
			const loadingUserDetailsQuery = { ...defaultUserDetailsQuery, loading: true };
			const wrapper = shallowComponent(loadingUserDetailsQuery);
			expect(wrapper).toMatchSnapshot();
		});
		it('should load TimeEntry if not loading and if authenticated', () => {
			const wrapper = shallowComponent(defaultUserDetailsQuery);
			wrapper.setState({ authenticated: true });
			expect(wrapper).toMatchSnapshot();
		});
		it('should load Login if not loading and not authenticated', () => {
			const wrapper = shallowComponent(defaultUserDetailsQuery);
			wrapper.setState({ authenticated: false });
			expect(wrapper).toMatchSnapshot();
		});
	});
	describe('componentWillReceiveProps', () => {
		it('should authenticate when props are correct', () => {
			const loadingQuery = { ...defaultUserDetailsQuery, loading: true };
			const wrapper = shallowComponent(loadingQuery);

			token.getAuthToken = jest.fn(() => ('loggedIn'));
			wrapper.setProps({ userDetailsQuery: defaultUserDetailsQuery });

			expect(wrapper.state('authenticated')).toEqual(true);
		});
		it('should authenticate when props are not correct', () => {
			const loadingQuery = { ...defaultUserDetailsQuery, loading: true };
			const wrapper = shallowComponent(loadingQuery);

			const errorQuery = { ...defaultUserDetailsQuery, error: true };
			wrapper.setProps({ userDetailsQuery: errorQuery });

			expect(wrapper.state('authenticated')).toEqual(false);
		});
	});
});
