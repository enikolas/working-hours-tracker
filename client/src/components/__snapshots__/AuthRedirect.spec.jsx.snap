// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthRedirect render should load Login if not loading and not authenticated 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthRedirect
    addTimeEntryMutation={[Function]}
    signIn={Object {}}
    updateTimeEntryMutation={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": false,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "signIn": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "signIn": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`AuthRedirect render should load TimeEntry if not loading and if authenticated 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthRedirect
    addTimeEntryMutation={[Function]}
    signIn={Object {}}
    updateTimeEntryMutation={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": false,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "addTimeEntryMutation": [Function],
      "allEntriesQuery": Object {
        "allEntries": Object {
          "admission": "",
          "entries": Array [],
          "name": "",
        },
        "error": "",
        "loading": false,
      },
      "updateTimeEntryMutation": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "addTimeEntryMutation": [Function],
        "allEntriesQuery": Object {
          "allEntries": Object {
            "admission": "",
            "entries": Array [],
            "name": "",
          },
          "error": "",
          "loading": false,
        },
        "updateTimeEntryMutation": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`AuthRedirect render should render page loading if loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthRedirect
    addTimeEntryMutation={[Function]}
    signIn={Object {}}
    updateTimeEntryMutation={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": true,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "active": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "active": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
